src/ecco/lm.py

```diff
@@ -276,25 +204,81 @@ def generate(self, input_str: str,
         if self.verbose:
             viz_id = self.display_input_sequence(input_ids[0])
 
-        while cur_len < max_length:
-            output_token_id, output = self._generate_token(encoder_input_ids=input_ids,
-                                                           encoder_attention_mask=attention_mask,
-                                                           decoder_input_ids=decoder_input_ids,
-                                                           past=past,  # Note: this is not currently used
-                                                           temperature=temperature,
-                                                           top_k=top_k,
-                                                           top_p=top_p,
-                                                           do_sample=do_sample,
-                                                           attribution_flags=attribution)
+        # Get model output
+        self._remove_hooks() # deactivate hooks: we will run them for the last model forward only
+        output = self.model.generate(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            num_beams=beam_size,
+            # FIXME: +1 in max_length to account for first start token in decoder, find a better way to do this
+            max_length=(generate or max_length - cur_len) + 1 if self.model_type == 'enc-dec' else max_length,
+            do_sample=do_sample,
+            top_p=top_p,
+            top_k=top_k,
+            temperature=temperature,
+            return_dict_in_generate=True,
+            output_scores=True,
+            **generate_kwargs
+        )
+
+        # Get prediction logits for each chosen prediction id
+        prediction_logits, prediction_ids = [], []
+        if output.__class__.__name__.endswith("EncoderDecoderOutput"):
+            prediction_ids, prediction_scores = output.sequences[0][1:], output.scores
+
+        elif output.__class__.__name__.endswith("DecoderOnlyOutput"):
+            prediction_ids, prediction_scores = output.sequences[0][n_input_tokens:], output.scores
+
```

Comment:
> Another great PR! Looks good to me!

Thanks! Always a pleasure to contribute 💪 

> My understanding is that output is of type ModelOutput or its subclasses (either class transformers.modeling_outputs.Seq2SeqModelOutput or class transformers.modeling_outputs.CausalLMOutput?) Is that correct?

That's when the models are called directly with `model(input)`. When using the `generate` method, the possible output types are defined [here](https://huggingface.co/docs/transformers/master/en/main_classes/model#transformers.generation_utils.GenerationMixin.generate.returns). 

In another note, the flag that they mention ( `model.config.is_encoder_decoder`) is pretty useful. If only there was a way to identify causal LMs such as GPT we would not need the `type` attribute in our model configs. However, causal LMs are a very special case in this `transformers` package because they have `model.config.is_encoder_decoder=False` but they act as encoder decoders 😄